memcached/memcached.go

```diff
@@ -23,47 +63,80 @@ func NewMemCachedClient(serverList []string) (*CacheClient, error) {
+// time from now (up to 1 month), or an absolute Unix epoch time.
+// Zero means the Item has no expiration time.
+// It returns (nil, err) if there's any other error, else returns an interface{} object.
+func (c *CacheClient) GetItem(key string, expiration int32, dbCallBack func() (interface{}, error)) (interface{}, error) {
 	item, err := c.client.Get(key)
+	if err != nil {
+		if err == memcache.ErrCacheMiss {
+			value, err := dbCallBack()
+			if err != nil {
+				return nil, err
+			}
+			_, err = c.AddItem(key, value, expiration)
+			if err != nil {
+				return nil, err
+			}
+			return value, nil
+		}
+		return nil, err
+	}
+	res, err := BytesToEmptyInterface(item.Value)
 	if err != nil {
 		return nil, err
 	}
-	return item, nil
+	return res, nil
 }
 
-// saveItem saves an Item to cache.
-// It returns error if it is unable to save the Item.
-func (c *CacheClient) saveItem(item *memcache.Item) error {
-	err := c.client.Add(item)
+// AddItem saves an Item to cache.
+// It returns false,error if it is unable to save the Item.
+// expiration is the cache expiration time, in seconds: either a relative
+// time from now (up to 1 month), or an absolute Unix epoch time.
+// Zero means the Item has no expiration time.
+func (c *CacheClient) AddItem(key string, value interface{}, expiration int32) (bool, error) {
+	item, err := CreateMemCacheObject(key, value, expiration)
+	if err != nil {
+		return false, err
+	}
+	err = c.client.Add(item)
 	if err != nil {
-		return err
+		return false, err
 	}
-	return nil
+	return true, nil
 }
 
-// updateItem updates an Item in cache. If adds the item if the key doesn't exist
+// UpdateItem updates an Item in cache. If addIfNotExists is true, the item is added if the key doesn't exist.
 // It returns error if it is unable to update the Item.
-func (c *CacheClient) updateItem(item *memcache.Item) error {
-	err := c.client.Replace(item)
+// expiration is the cache expiration time, in seconds: either a relative
+// time from now (up to 1 month), or an absolute Unix epoch time.
+// Zero means the Item has no expiration time.
+func (c *CacheClient) UpdateItem(key string, value interface{}, expiration int32, addIfNotExists bool) (bool, error) {
+	item, err := CreateMemCacheObject(key, value, expiration)
+	if err != nil {
+		return false, err
+	}
+	err = c.client.Replace(item)
 	if err != nil {
 		//unable to find key in cache
-		err = c.saveItem(item)
-		if err != nil {
-			return err
+		if addIfNotExists {
+			return c.AddItem(key, value, expiration)
 		}
-		return nil
+		return false, err
 	}
-	return nil
+	return true, nil
 }
 
-// deleteItem deletes a given key from the server.
-// It returns error if delete was unsuccessful.
-func (c *CacheClient) deleteItem(key string) error {
+// DeleteItem deletes a given key from the server.
+// It returns false,error if delete was unsuccessful.
+func (c *CacheClient) DeleteItem(key string) (bool, error) {
```

Comment:
add deleteWithDelay function which would add the old object back with a 5 mins delay